File: ArcadiaGames/app/src/main/java/com/example/arcadiagames/FragmentCatalogo.java
package com.example.arcadiagames;

import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import com.google.android.material.search.SearchView;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.google.android.material.chip.Chip;
import com.google.android.material.chip.ChipGroup;
import com.google.android.material.search.SearchBar; 

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class FragmentCatalogo extends Fragment {

    private RecyclerView recyclerView;
    private JuegoAdapter adapter;
    private List<Juego> listaCompleta;
    private SearchBar searchBar;
    private ChipGroup chipGroup;
    private SearchView searchView;

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_catalogo, container, false);

        // Inicializar vistas
        recyclerView = view.findViewById(R.id.rv_catalogo_completo);
        searchBar = view.findViewById(R.id.search_bar_catalogo);
        searchView = view.findViewById(R.id.search_view_catalogo);
        chipGroup = view.findViewById(R.id.chip_group_categorias);

        // 1. Crear la lista de datos (Mockup)
        llenarListaProvisional();

        // 2. Configurar RecyclerView
        recyclerView.setLayoutManager(new LinearLayoutManager(getContext()));
        adapter = new JuegoAdapter(new ArrayList<>(listaCompleta));
        recyclerView.setAdapter(adapter);

        // 3. Configurar Filtros
        configurarFiltros();
        searchView.setupWithSearchBar(searchBar);

        return view;
    }

    private void configurarFiltros() {
        // Escuchar cuando el usuario escribe en el SearchView
        searchView.getEditText().addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {}

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                // Actualizamos el texto de la SearchBar para que se vea lo que escribimos al cerrar
                searchBar.setText(s.toString());
                filtrar(s.toString());
            }

            @Override
            public void afterTextChanged(Editable s) {}
        });

        // Escuchar cambios en los Chips
        chipGroup.setOnCheckedStateChangeListener((group, checkedIds) -> {
            // Obtenemos el texto actual del SearchView para no perder el filtro de nombre
            String textoBusqueda = searchView.getText().toString();
            filtrar(textoBusqueda);
        });
    }

    private void filtrar(String texto) {
        List<Juego> listaFiltrada = new ArrayList<>();

        // 1. Obtener Tags seleccionados
        List<String> etiquetasSeleccionadas = new ArrayList<>();
        for (int id : chipGroup.getCheckedChipIds()) {
            Chip chip = chipGroup.findViewById(id);
            if (chip != null) {
                etiquetasSeleccionadas.add(chip.getText().toString());
            }
        }

        // 2. Lógica de filtrado combinada
        for (Juego juego : listaCompleta) {
            boolean coincideNombre = juego.getNombre().toLowerCase().contains(texto.toLowerCase());

            boolean coincideChip = etiquetasSeleccionadas.isEmpty();
            if (!etiquetasSeleccionadas.isEmpty()) {
                for (String tag : juego.getTags()) {
                    if (etiquetasSeleccionadas.contains(tag)) {
                        coincideChip = true;
                        break;
                    }
                }
            }

            if (coincideNombre && coincideChip) {
                listaFiltrada.add(juego);
            }
        }

        // 3. Actualizar el adaptador
        adapter.setFilteredList(listaFiltrada);
    }

    private void llenarListaProvisional() {
        listaCompleta = new ArrayList<>();
        // Ejemplo de datos con los nuevos nombres
        listaCompleta.add(new Juego("Super Morio World", "C morió.", R.drawable.morio, "4000.99€", Arrays.asList("MeloInvento")));
        listaCompleta.add(new Juego("Poly Racing", "Carreras poligonales. Mira un vértice", R.drawable.polyracing, "0.09€", Arrays.asList("PolyStation")));
        listaCompleta.add(new Juego("Xbob El con-Xtructor", "Construye como nadie", R.drawable.wicher3, "49.99€", Arrays.asList("Xbob")));
        listaCompleta.add(new Juego("PC Master Race", "Racist era algo de carreras ¿no?.", R.drawable.pcmasterrace, "50.00€", Arrays.asList("PC")));
        listaCompleta.add(new Juego("Multi Plataforma", "Funciona en todo.", R.drawable.wicher3, "3.99€", Arrays.asList("PolyStation", "PC", "Xbob", "MeloInvento")));
        listaCompleta.add(new Juego("Facto Horio", "Tira factos en el horio.", R.drawable.wicher3, "39.99€", Arrays.asList("PolyStation", "PC")));

    }
}

File: ArcadiaGames/app/src/main/java/com/example/prueba1integrador/FragmentCatalogo.kt
package com.example.prueba1integrador

import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.LinearLayoutManager
import com.example.prueba1integrador.databinding.FragmentCatalogoBinding
import com.google.android.material.chip.Chip
import android.content.Intent

class FragmentCatalogo : Fragment() {

    private var _binding: FragmentCatalogoBinding? = null
    private val binding get() = _binding!!

    private lateinit var adapter: JuegoAdapter
    private var listaCompleta: List<Juego> = emptyList()
    private val inventoryManager = FirebaseInventoryManager()

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        _binding = FragmentCatalogoBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        setupRecyclerView()
        setupData()
        setupFilters()
        setupFab()
    }

    private fun setupData() {
        inventoryManager.consultarInventario(object :
            FirebaseInventoryManager.InventoryCallback {
            override fun onDataLoaded(lista: List<Juego>) {
                listaCompleta = lista
                adapter.setFilteredList(listaCompleta)
            }
        })
    }

    private fun setupFab() {
        val rolUsuario = activity?.intent?.getStringExtra("ROL_USUARIO") ?: "cliente"

        if (rolUsuario == "admin") {
            binding.fabAgregarJuego.visibility = View.VISIBLE
            binding.fabAgregarJuego.setOnClickListener {
                startActivity(Intent(requireContext(), AnadirProductoActivity::class.java))
            }
        } else {
            binding.fabAgregarJuego.visibility = View.GONE
        }
    }

    private fun setupRecyclerView() {
        binding.rvCatalogoCompleto.layoutManager = LinearLayoutManager(context)

        val rolUsuario = activity?.intent?.getStringExtra("ROL_USUARIO") ?: "cliente"
        val esAdmin = (rolUsuario == "admin")

        adapter = JuegoAdapter(emptyList(), esCarousel = false, esAdmin = esAdmin)
        binding.rvCatalogoCompleto.adapter = adapter
    }

    private fun setupFilters() {

        binding.searchViewCatalogo.setupWithSearchBar(binding.searchBarCatalogo)

        binding.searchViewCatalogo.editText.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {}

            override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {
                val texto = s.toString()
                binding.searchBarCatalogo.setText(texto)
                filtrar(texto)
            }

            override fun afterTextChanged(s: Editable?) {}
        })

        // Ahora solo hay un chip seleccionado
        binding.chipGroupCategorias.setOnCheckedStateChangeListener { group, checkedIds ->
            filtrar(binding.searchViewCatalogo.text.toString())
        }
    }

    private fun filtrar(texto: String) {

        val query = texto.lowercase()

        // 🔥 Como ahora es singleSelection, solo puede haber 1
        val selectedId = binding.chipGroupCategorias.checkedChipId

        val plataformaSeleccionada = if (selectedId != View.NO_ID) {
            val chip = binding.chipGroupCategorias.findViewById<Chip>(selectedId)
            chip.text.toString().lowercase()
        } else {
            null
        }

        val listaFiltrada = listaCompleta.filter { juego ->

            val coincideNombre = juego.nombre.lowercase().contains(query)

            val coincideChip = plataformaSeleccionada == null ||
                    juego.categoria.lowercase().contains(plataformaSeleccionada) ||
                    juego.plataforma.lowercase().contains(plataformaSeleccionada)

            coincideNombre && coincideChip
        }

        adapter.setFilteredList(listaFiltrada)
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }
}
